### ISSUE TITLE
panic while attempting key-change if cloned snapshots of an encrypted dataset are present

### ISSUE THREAD

**Issue by jasonrm** [zfsonlinux/zfs#6703]
### System information
Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | Arch Linux
Distribution Version    | N/A
Linux Kernel                 | 4.13.3-1-ARCH
Architecture                 | x86_64
ZFS Version                  |  0.7.0-98_g01ff0d754
SPL Version                  |  0.7.0-13_ge8474f9

### Describe the problem you're observing

It appears to me that after an encrypted dataset is cloned, attempting to use `zfs change-key` results in a panic.

### Describe how to reproduce the problem

__Panic: keyformat=passphrase -> keyformat=raw__
```bash
fallocate -l 512M /tmp/throwaway
dd if=/dev/urandom of=/tmp/throwaway.key bs=32 count=1
echo 'C74AE2525459' | zpool create -O encryption=on -O keyformat=passphrase -O mountpoint=none throwaway /tmp/throwaway
zfs create throwaway/docker
zfs create throwaway/docker/image
zfs snapshot throwaway/docker/image@snap
zfs clone throwaway/docker/image@snap throwaway/docker/cloned
zfs get -t all -r encryption,encryptionroot,keyformat,keylocation throwaway/docker
zfs change-key -o keyformat=raw -o keylocation=file:///tmp/throwaway.key throwaway
```

__Panic: keyformat=passphrase -> keyformat=passphrase__
```bash
fallocate -l 512M /tmp/throwaway
echo 'C74AE2525459' | zpool create -O encryption=on -O keyformat=passphrase -O mountpoint=none throwaway /tmp/throwaway
zfs create throwaway/docker
zfs create throwaway/docker/image
zfs snapshot throwaway/docker/image@snap
zfs clone throwaway/docker/image@snap throwaway/docker/cloned
zfs get -t all -r encryption,encryptionroot,keyformat,keylocation throwaway/docker
echo 'F81BF4778069' | zfs change-key throwaway
```

__Panic: keyformat=raw -> keyformat=raw__
```bash
fallocate -l 512M /tmp/throwaway
dd if=/dev/urandom of=/tmp/throwaway.key bs=32 count=1
zpool create -O encryption=on -O keyformat=raw -O keylocation=file:///tmp/throwaway.key -O mountpoint=none throwaway /tmp/throwaway
zfs create throwaway/docker
zfs create throwaway/docker/image
zfs snapshot throwaway/docker/image@snap
zfs clone throwaway/docker/image@snap throwaway/docker/cloned
zfs get -t all -r encryption,encryptionroot,keyformat,keylocation throwaway/docker
dd if=/dev/urandom of=/tmp/throwaway.key bs=32 count=1
zfs change-key -o keyformat=raw -o keylocation=file:///tmp/throwaway.key throwaway
```

__Panic: change-key with no options?__
I'm not sure if this is the same as trying to switch from keyformat=raw -> keyformat=passphrase

```bash
fallocate -l 512M /tmp/throwaway
dd if=/dev/urandom of=/tmp/throwaway.key bs=32 count=1
zpool create -O encryption=on -O keyformat=raw -O keylocation=file:///tmp/throwaway.key -O mountpoint=none throwaway /tmp/throwaway
zfs create throwaway/docker
zfs create throwaway/docker/image
zfs snapshot throwaway/docker/image@snap
zfs clone throwaway/docker/image@snap throwaway/docker/cloned
zfs get -t all -r encryption,encryptionroot,keyformat,keylocation throwaway/docker
zfs change-key
```

### Include any warning/errors/backtraces from the system logs

All of the commands above result in the same panic.

```
$ dmesg -w
VERIFY3(0 == spa_keystore_dsl_key_hold_dd(dp->dp_spa, dd, ((char *)__func__), &dck)) failed (0 == 13)
PANIC at dsl_crypt.c:1347:spa_keystore_change_key_sync_impl()
Showing stack for process 3553
CPU: 2 PID: 3553 Comm: txg_sync Tainted: P           O    4.13.3-1-ARCH #1
Hardware name: HP ProLiant MicroServer Gen8, BIOS J06 11/02/2015
Call Trace:
 dump_stack+0x63/0x8b
 spl_dumpstack+0x42/0x50 [spl]
 spl_panic+0xc8/0x110 [spl]
 ? spa_keystore_dsl_key_hold_dd.isra.13+0x1ab/0x1c0 [zfs]
 spa_keystore_change_key_sync_impl+0x3a2/0x440 [zfs]
 spa_keystore_change_key_sync_impl+0x193/0x440 [zfs]
 spa_keystore_change_key_sync_impl+0x193/0x440 [zfs]
 ? zap_lookup+0x16/0x20 [zfs]
 spa_keystore_change_key_sync+0x1aa/0x3e0 [zfs]
 ? dmu_buf_rele+0x36/0x40 [zfs]
 ? dsl_dir_rele+0x31/0x40 [zfs]
 ? spa_keystore_change_key_check+0x267/0x4e0 [zfs]
 dsl_sync_task_sync+0xa6/0x110 [zfs]
 dsl_pool_sync+0x300/0x430 [zfs]
 spa_sync+0x43f/0xd80 [zfs]
 txg_sync_thread+0x2d2/0x4a0 [zfs]
 ? finish_task_switch+0x75/0x200
 ? txg_delay+0x1b0/0x1b0 [zfs]
 thread_generic_wrapper+0x72/0x80 [spl]
 kthread+0x125/0x140
 ? __thread_exit+0x20/0x20 [spl]
 ? kthread_create_on_node+0x70/0x70
 ret_from_fork+0x25/0x30
```

**Comment by tcaputi** [comment]
Thanks for the easy steps to reproduce. i'll get this fixed first thing tomorrow morning.

**Comment by tcaputi** [comment]
This is a great catch. I have a fix and will post it to #6595 as soon as I get the chance to run it through the existing regression tests (at latest tomorrow morning).

**Comment by jasonrm** [comment]
Your testing was likely much more extensive than mine, but wanted to confirm that 060e4cf worked for me.

Sincerely appreciate all the work you (along with others, of course) have done to get encryption into master. Glad I was finally able to help in some minor way.
